### ISSUE TITLE
Unable to compile all protofiles in directory tree with `protoc`.

### ISSUE THREAD

**Issue by nmccrea** [protocolbuffers/protobuf#11656]
**Environment**
Protobufs Version: 3.21.9 (Installed and compiled from Conan Center)
Operating System: MacOS Ventura 13.1
Shell: zsh

Apple clang version 14.0.0 (clang-1400.0.29.202)
Target: arm64-apple-darwin22.2.0
Thread model: posix

**Problem Statement**
Passing recursive directory expansion (aka "globbing" or "globstar" or "wildcard") to `protoc` is not working for me. Furthermore it does not match the globbing behavior of my shell. As a result I am unable to write a `protoc` script that will not require maintenance as the directory structure changes.

[This StackOverflow post](https://stackoverflow.com/questions/59314322/protobuf-build-protoc-select-all-proto-files-in-a-directory) seems to be demonstrating the same problem, and suggests this is a system-specific problem, as one of the respondants is experiencing the desired behavior. But the post does not provide resolution. The comments suggest it might be due to differing shell behaviors, but as demonstrated below my native shell behavior is as expected.

**Example Setup**
My protofiles are in a structure like this:

```
protofiles/
  cheese.proto
  crackers.proto
  subdirectory/
    beans.proto
    cornbread.proto
    extrasubdirectory/
      pickles.proto
```

**Test of Native Shell Behavior**

```bash
$ find protofiles/**/*.proto
protofiles/cheese.proto
protofiles/crackers.proto
protofiles/subdirectory/beans.proto
protofiles/subdirectory/cornbread.proto
protofiles/subdirectory/extrasubdirectory/pickles.proto
```

**Results of `protoc` Command**

This is the command I would like to use:

```bash
$ protoc -I=protofiles --cpp_out=src protofiles/**/*.proto

$ find src/**  
src/subdirectory
src/subdirectory/cornbread.pb.h
src/subdirectory/cornbread.pb.cc
src/subdirectory/beans.pb.cc
src/subdirectory/beans.pb.h
```

As you can see it is treating the globstar `**` as a "exactly one directory level". In order to compile the top-level directory, I must run:

```bash
$ protoc -I=protofiles --cpp_out=src protofiles/*.proto

$ find src/**
src/cheese.pb.cc
src/cheese.pb.h
src/crackers.pb.cc
src/crackers.pb.h
```

And to get the bottom level directory:

```bash
$ protoc -I=protofiles --cpp_out=src protofiles/**/**/*.proto

$ find src/**
src/subdirectory
src/subdirectory/extrasubdirectory
src/subdirectory/extrasubdirectory/pickles.pb.h
src/subdirectory/extrasubdirectory/pickles.pb.cc
```

The only way to get my whole tree to compile is to manually specify as many levels of the tree as there are, which is messy and brittle:

```bash
$ protoc -I=protofiles --cpp_out=src protofiles/*.proto protofiles/**/*.proto protofiles/**/**/*.proto

$ find src/**
src/cheese.pb.cc
src/cheese.pb.h
src/crackers.pb.cc
src/crackers.pb.h
src/subdirectory
src/subdirectory/cornbread.pb.h
src/subdirectory/cornbread.pb.cc
src/subdirectory/extrasubdirectory
src/subdirectory/extrasubdirectory/pickles.pb.h
src/subdirectory/extrasubdirectory/pickles.pb.cc
src/subdirectory/beans.pb.cc
src/subdirectory/beans.pb.h
```

**Comment by shaod2** [comment]
Instead of sticking with `**`, I was thinking about attaching the results of `find` into protoc? Something like `protoc -I=protofiles --cpp_out=src $(find protofiles/**/*.proto)`.

**Comment by nmccrea** [comment]
Hi @shaod2. This was good thinking and has helped me identify that `protoc` is not actually the source of my problem.

I didn't mention it in my original post because I did not think it was relevant, but I am running the `protoc` command from a Python function in my Conanfile. I naively assumed this function would simply be passing my command string to the shell unaltered but it appears either Conan or Python is changing the behavior of the globstar before it reaches my shell. (Pretty abysmal behavior in my humble opinion.)

I discovered this because using the method you described resulted in the same behavior as in my original post, which was very surprising since we know that the `find` command is returning all the desired files. Indeed hand-writing the files returned by `find` into the `protoc` command works as desired, so it's clear something else is going on before my command is reaching the shell. I've verified this with a couple of other simple tests.

I will close this issue. Thanks very much for your help.
